src/app/clusters/occupancy-sensor-server/OccupancySensingCluster.cpp

```diff
@@ -0,0 +1,333 @@
+    default:
+        return Protocols::InteractionModel::Status::UnsupportedAttribute;
+    }
+}
+
+DataModel::ActionReturnStatus OccupancySensingCluster::WriteAttribute(const DataModel::WriteAttributeRequest & request,
+                                                                      AttributeValueDecoder & aDecoder)
+{
+    VerifyOrDie(request.path.mClusterId == OccupancySensing::Id);
+
+    switch (request.path.mAttributeId)
+    {
+    case Attributes::HoldTime::Id:
+    case Attributes::PIROccupiedToUnoccupiedDelay::Id:
+    case Attributes::UltrasonicOccupiedToUnoccupiedDelay::Id:
+    case Attributes::PhysicalContactOccupiedToUnoccupiedDelay::Id: {
+        if (!mHasHoldTime)
+        {
+            return Protocols::InteractionModel::Status::UnsupportedAttribute;
+        }
+        uint16_t newHoldTime;
+        ReturnErrorOnFailure(aDecoder.Decode(newHoldTime));
+        return NotifyAttributeChangedIfSuccess(request.path.mAttributeId, SetHoldTime(newHoldTime));
+    }
+    default:
+        return Protocols::InteractionModel::Status::UnsupportedAttribute;
+    }
+}
+
+CHIP_ERROR OccupancySensingCluster::Attributes(const ConcreteClusterPath & clusterPath,
+                                               ReadOnlyBufferBuilder<DataModel::AttributeEntry> & builder)
+{
+    AttributeListBuilder listBuilder(builder);
+
+    const AttributeListBuilder::OptionalAttributeEntry optionalAttributes[] = {
+        { mHasHoldTime, Attributes::HoldTime::kMetadataEntry },
+        { mHasHoldTime, Attributes::HoldTimeLimits::kMetadataEntry },
+        { mHasHoldTime && mShowDeprecatedAttributes && mFeatureMap.Has(Feature::kPassiveInfrared),
+          Attributes::PIROccupiedToUnoccupiedDelay::kMetadataEntry },
+        { mHasHoldTime && mShowDeprecatedAttributes && mFeatureMap.Has(Feature::kUltrasonic),
+          Attributes::UltrasonicOccupiedToUnoccupiedDelay::kMetadataEntry },
+        { mHasHoldTime && mShowDeprecatedAttributes && mFeatureMap.Has(Feature::kPhysicalContact),
+          Attributes::PhysicalContactOccupiedToUnoccupiedDelay::kMetadataEntry },
+    };
+
+    return listBuilder.Append(Span(OccupancySensing::Attributes::kMandatoryMetadata), Span(optionalAttributes));
+}
+
+DataModel::ActionReturnStatus OccupancySensingCluster::SetHoldTime(uint16_t holdTime)
+{
+    VerifyOrReturnError(mHasHoldTime, Protocols::InteractionModel::Status::UnsupportedAttribute);
+    VerifyOrReturnError(holdTime >= mHoldTimeLimits.holdTimeMin, Protocols::InteractionModel::Status::ConstraintError);
+    VerifyOrReturnError(holdTime <= mHoldTimeLimits.holdTimeMax, Protocols::InteractionModel::Status::ConstraintError);
+
+    if (mHoldTime == holdTime)
+    {
+        return DataModel::ActionReturnStatus::FixedStatus::kWriteSuccessNoOp;
+    }
+
+    mHoldTime = holdTime;
+
+    // If a timer is currently active, we need to adjust its duration to reflect the new hold time.
+    if (mTimerDelegate && mTimerDelegate->IsTimerActive(this))
```

Comment:
It seems better if we do not allow null timer delegates - force the callers to provide one.

Reply:
Or maybe since this is tied to hold time, we could make the whole `if you have hold time you also have a timer delegate` inside a std::optional.